### ISSUE TITLE
Add Source Mode/Current Volume Level as Entity

### ISSUE THREAD

**Issue by aiden21c** [mvdwetering/yamaha_ynca#132]
I'm not sure if I have missed something, but I would love a way to set the source mode and volume level as an entity. This would allow me to better control a scene I might use such as "sleep" where I set the receiver to Airplay, set sleep timer to 60 and set volume to about 36.

**Comment by mvdwetering** [comment]
The source and volume are part of the media_player entity. 

From a quick experiment it seems like Home Assistant stores all settings that are part of the media_player entity in the scene.

**Comment by aiden21c** [comment]
Thank you for your response. I will give this another go in HA and see if I am able to get it to function as I'd like.

**Comment by mvdwetering** [comment]
That is weird, could you attach the Diagnostics file?
You can get it by opening the triple dot menu on the integration card and then select the option "Download diagnostics".

**Comment by aiden21c** [comment]
[config_entry-yamaha_ynca-92c87f729c25299388a189a17d0195cc.zip](https://github.com/mvdwetering/yamaha_ynca/files/10736854/config_entry-yamaha_ynca-92c87f729c25299388a189a17d0195cc.zip)
See attatched. Thank you for your quick reply

**Comment by mvdwetering** [comment]
I am seeing it when using your diagnostics data in my emulator.
Can you use the search and type "media" in the seach bar? I then just get the 2 media players.

Or instead of opening the entities view select the devices view from the Yamaha YNCA integration card and select the device for the Main zone. The device page you get then also shows the enties for that device like below, which shows the meidaplayer at the top of the Control section.

![image](https://user-images.githubusercontent.com/732514/219481593-74a7901e-3070-4f34-87d4-a91a3842f7b6.png)

**Comment by aiden21c** [comment]
I moved to the dev branch do so some testing regarding a prior issue thread, and it seems that everything is as you have said on this branch. I was certain however this was not the case on the production branch, with the media player integration missing for the main zone in the gui. Once I switch back to the production branch I will monitor and see if there are any further issues. Thanks for your help.

**Comment by mvdwetering** [comment]
There is a small chance that initializing of the media_player might fail for unmappable sources, but that should have resulted in an error in the Home Asssistant logs. If that is the case It would be nice to get that error, so I can make the code more robust for such a case.

**Comment by mikesalz** [comment]
@mvdwetering I am not seeing volume as a separate entity.  I can access it using 
`{{ state_attr('media_player.yamaha_rx_a6a', 'volume_level') | float }}`

But should it be an entity as well?

Also, I am wondering if there is a way to see and set the current Subwoofer Trim level?  

![image](https://github.com/mvdwetering/yamaha_ynca/assets/46855061/7d105a5d-37e4-44b8-a1ec-175882aa1165)

**Comment by mvdwetering** [comment]
The volume control is part of the mediaplayer entity. That is the intended way to do it in Home Assistant as far as I know.

I don't know about subwoofer trim, but we can try and see. 
Can you change that setting on your receiver to the minimum and maximum values and then download the device diagnostics from Home Assistant. Then attach the file here (you can drag and drop the file in the textfield). The diagnostics contains the last commands received from the receiver so if it sent something it is probably easy to figure out how it works.

![image](https://github.com/mvdwetering/yamaha_ynca/assets/732514/a66f8b7f-7f63-4e35-806d-6ace2f7e5f9a)

**Comment by mikesalz** [comment]
Thanks for responding so quickly!  I tried adjusting it several times, but I'm not seeing it in the log.  I even adjusted the volume and then the subwoofer trim again after that so I would see where it should fall in the log, but nothing.  Maybe you'll find something that I couldn't.  It's weird though.  You'd think it if was available in their web UI, it would be available in an API.

[config_entry-yamaha_ynca-33911dd573770ab4f062bcd48efeb91a.json (6).txt](https://github.com/mvdwetering/yamaha_ynca/files/12422983/config_entry-yamaha_ynca-33911dd573770ab4f062bcd48efeb91a.json.6.txt)

**Comment by mikesalz** [comment]
@mvdwetering I doubt this will be helpful to you, but mentioning it just in case.  I opened the Chrome developer tools and then tested some changes in the Yamaha Web Setup.

Changing the volume to -39.0dB yielded a payload that looks like this:
`<YAMAHA_AV cmd="PUT"><Main_Zone><Volume><Lvl><Val>-390</Val><Exp>1</Exp><Unit>dB</Unit></Lvl></Volume></Main_Zone></YAMAHA_AV>`

I then bumped the Subwoofer Trim to +1.0dB and got a payload of this:
`<YAMAHA_AV cmd="PUT"><Main_Zone><Volume><Subwoofer_Trim><Val>10</Val><Exp>1</Exp><Unit>dB</Unit></Subwoofer_Trim></Volume></Main_Zone></YAMAHA_AV>`

I'm not sure if that is at all helpful to you, but I'm throwing it out there just in case!

**Comment by mvdwetering** [comment]
>You'd think it if was available in their web UI, it would be available in an API.

Indeed, but Yamaha keeps multiple APIs on their receivers for some reason and they all seem to have slightly different feature sets.


That XML protocol looks like the one that is used by the standard Yamaha integration in Home Assistant, which it totally different from the protocol used by this integration. But I have the impression that that integration is not really being updated as it is still yaml only.

Did you try the Yamaha MusicCast integration in Home Assistant? 
Not sure what features it offers since my receiver is too old, but maybe has something.

You could try to put up a feature request with Home Assistant to add support for either of the official integrations.


btw I looked around a bit and found a mention of this command `@MAIN:SWFRTRIM=?` which looks like it could be SubWooFeR trim. Not sure for which receiver it is. We could try and see how the receiver responds.

To send the command go to the developer tools in Home Assistant and select the "Yamaha (YNCA): Send raw YNCA command" service.
In the targets field select your receiver (any zone is fine)
In the raw data field input `@MAIN:SWFRTRIM=?`
And then call the service (make sure the receiver is On before sending the command)

After this get a new diagnostics file and look at the end. You should see the command being sent and the response. It could be a response with the current value or an error like UNDEFINED or RESTRICTED. Given the first experiment I would expect an error, but does not hurt to try.

**Comment by mikesalz** [comment]
Welp, no luck with SWFRTRIM.  But I appreciate your help!  I had previously ignored the MusicCast integration because I assumed it was only for streaming media.  But per your suggested I gave it a try just now.  And guess what - It does have an entity for Subwoofer Trim (they call it subwoofer volume)!  I guess I'll need to keep both integrations now - Yours to control scene selection and MusicCast to control the sub volume.

![image](https://github.com/mvdwetering/yamaha_ynca/assets/46855061/4e6dbf9e-c90f-4ac6-ac06-e8870cfd2a19)
